### PR TITLE
Reference region cleanup

### LINKED ISSUE #1285: Clarify strandedness in ReferenceRegion apply methods
Given the discussion by me @akmorrow13 and @jpdna in https://github.com/bigdatagenomics/utils/pull/93#issuecomment-261802968, we should probably clarify this. My proposal is:

* No `apply` methods for datatypes that can be stranded (`AlignmentRecord`s and `Feature`s?)
* Instead, two methods:
  * `def independent` which returns a reference region with an independent strand
  * `def stranded` which returns a reference region with orientation set

We'd then want to modify something in `GenomicRDD` to clarify whether the data in the RDD is stranded or not, for data that can be strand (in)dependent.

I'll take a quick hack at this tomorrow; should be a quick change. If everyone thinks it looks good, we can get it in 0.21.0, else we can push it.

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #1285 and #1286. The first two commits need to be squashed together.

### CODE REVIEW

**Comment by fnothaft** [comment]
Ping @heuermh @jpdna for review.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 1]
```diff
@@ -220,17 +255,45 @@ object ReferenceRegion {
...
+   * @see stranded
    */
-  def apply(feature: Feature): ReferenceRegion = {
+  def unstranded(feature: Feature): ReferenceRegion = {
+    checkFeature(feature)
     new ReferenceRegion(feature.getContigName, feature.getStart, feature.getEnd)
   }
 
   /**
+   * Builds a reference region for a feature without strand information.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: `without strand information.` &rarr; `with strand set.`

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 2]
```diff
@@ -220,17 +255,45 @@ object ReferenceRegion {
...
+   *
+   * @param feature Feature to extract ReferenceRegion from.
+   * @return Extracted ReferenceRegion
+   *
+   * @see unstranded
+   */
+  def stranded(feature: Feature): ReferenceRegion = {
+    checkFeature(feature)
+    require(feature.getStrand != null,
+      "Strand is not defined in feature %s.".format(feature))
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: Add @throws doc for this

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 3]
```diff
@@ -148,7 +148,9 @@ object ReferenceRegion {
    * @return The site where this genotype lives.
    */
   def apply(genotype: Genotype): ReferenceRegion = {
-    ReferenceRegion(genotype.getVariant)
+    ReferenceRegion(genotype.getContigName,
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: A few general questions (not about this line specifically):

Rename `orientation` to `strand`?
Add strand parameters to other methods (e.g. `fromStart`, `toEnd`, `all`)?
Set strand when copying from one `ReferenceRegion` to another (i.e. line 234)?

**Comment by fnothaft** [comment]
Thanks for the comments @heuermh. I've rebased and pushed two commits addressing your comments. When this looks good to you, I'll squash commits 2–4 down into a single commit for merge.

**Comment by fnothaft** [comment]
Squashed down.

**Comment by heuermh** [comment]
Thank you, @fnothaft

**Pull request merged**